Issue #9631: Time/Interval Hours

hawkfish opened the pull request:
Change the time parsing utilities to support:
* Intervals greater than 24 hours
* Times of 24:00:00 (PG allows this)

fixes: #9631
fixes: duckdblabs/duckdb-internal#672

hawkfish committed:
Issue #9631: Time/Interval Hours

Change the time parsing utilities to support:
* Intervals greater than 24 hours
* Times of 24:00:00 (PG allows this)

fixes: #9631
fixes: duckdblabs/duckdb-internal#672

Mytherin commented on test/sql/types/interval/test_interval.test:
```diff
@@ -98,6 +98,12 @@ select interval '-05:12:34.567890' as test_interval;
 
+# Issue #9631 - Parse hms > 24:00:00
+query I
+SELECT ('35:10:00'::interval);
```
What about triple digit numbers/numbers that would overflow?

hawkfish replied:
Turns out that we can't overflow time because hours are 32 bit and time is 64 bit. So limiting to 10 digits solves both problems.

hawkfish committed:
Merge branch 'feature' into interval-time

hawkfish committed:
Issue #9631: Wide Interval Hours

PR Feedback: Extend the allowed number of hour digits
for intervals to the maximum of 9.

Tishj commented:
I'm looking at some Arrow stuff, and while testing around I made this:
```py
import pyarrow as pa
import duckdb

ARRAY_SIZE = 2**17+1

col1 = [i for i in range(0, ARRAY_SIZE)]
# "a" in the struct matches the value for col1
col2 = [{"a": i} for i in col1]
arrow_table = pa.Table.from_pydict(
	{"col1": col1},
	schema=pa.schema([("col1", pa.time32('s'))]),
)

res = duckdb.sql(
	f"""
	SELECT
		col1
	FROM arrow_table offset {ARRAY_SIZE-1}
"""
).show()
```

On 0.9.1:
```
┌──────────┐
│   col1   │
│   time   │
├──────────┤
│ 36:24:32 │
└──────────┘
```

On main:
```
duckdb.duckdb.InternalException: INTERNAL Error: Assertion triggered in file "/Users/thijs/DuckDBLabs/duckdb/src/common/types/time.cpp" on line 327: Time::IsValidTime(hour, min, sec, micros)
```

Tishj commented:
> Are you sure 0.9.1 doesn't run into the same assertion? Note that assertions are disabled in the release.

Aha, yes then probably 0.9.1 is also broken, I just ran a pip installed 0.9.1 which is release

hawkfish committed:
Merge branch 'feature' into interval-time

hawkfish commented:
Some more fallout coming - we have to change the max value of `TIME` to `24:00:00`, which changes the storage version db _contents_ (but not the format). Plus all tests that use `test_all_types()` need to be updated.

hawkfish committed:
Issue #9631: TIME Maximum Value

We had incorrectly set this as 23:59:59.999999
instead of 24:00:00 (as per PG)

hawkfish committed:
Issue #9631: TIME Maximum Value

Fix up some tests in environments that can't support `24:00:00`.

hawkfish committed:
Issue #9631: TIME Maximum Value

More test fixing.

hawkfish committed:
Issue #9631: TIME Maximum Value

Even more test fixing.

hawkfish committed:
Issue #9631: TIME Maximum Value

Remove extraneous quotes.

hawkfish committed:
Issue #9631: TIME Maximum Value

Fix return type.

Tishj commented on src/common/types/time.cpp:
```diff
@@ -301,7 +304,7 @@ dtime_t Time::FromTime(int32_t hour, int32_t minute, int32_t second, int32_t mic
 bool Time::IsValidTime(int32_t hour, int32_t minute, int32_t second, int32_t microseconds) {
 	if (hour < 0 || hour >= 24) {
-		return false;
+		return (hour == 24) && (minute == 0) && (second == 0) && (microseconds == 0);
```
Why are we limiting the TIME at 24 hours?
I thought we were allowing arbitrary time values in both TIME and INTERVAL?

hawkfish replied:
`TIME` is limited to `24:00:00` See https://www.postgresql.org/docs/current/datatype-datetime.html

Tishj replied:
The issue I linked that I discovered in our Arrow scan is that Arrow does support > 24 hour time values, I don't know the limit but I got it to produce 36 hours, could be much higher.

Tishj replied:
It does seem like Postgres hard caps at 24 hours though.
```
postgres=# select CAST('24:00:00' AS time);
   time   
----------
 24:00:00
(1 row)

postgres=# select CAST('36:00:00' AS time);
ERROR:  date/time field value out of range: "36:00:00"
LINE 1: select CAST('36:00:00' AS time);
                    ^
postgres=# select CAST('24:00:01' AS time);
ERROR:  date/time field value out of range: "24:00:01"
LINE 1: select CAST('24:00:01' AS time);
```

Does that mean we need to actively check scanned time values from Arrow to see if they break our accepted bounds?

hawkfish replied:
Ah. Well we used to be restricted to `23:59:59.999999` so it's a little better! But yes reading back could be an issue. But this is about PG compatibility.

hawkfish replied:
yes I think so.

> Does that mean we need to actively check scanned time values from Arrow to see if they break our accepted bounds?

Sadly I think so. Or maybe map them to intervals?

Mytherin commented:
Could you resolve the merge conflicts?

hawkfish committed:
Merge branch 'main' into interval-time

hawkfish committed:
Merge branch 'main' into interval-time

Mytherin commented:
Thanks!
